### PR TITLE
Make JS object->Hash conversion traverse arrays

### LINKED ISSUE #1095: Hash.new with JS object w/ arrays as values doesn't recurse into the array
I noticed it because we're using `gon` to transfer state to our client apps. When I ran this on the opal side to get my initial state:

```ruby
Hash.new(`gon`)
```

The top-level values were good, and so were any nested hashes, except ones contained within arrays. For example ([try link](http://opalrb.org/try/?code:gon%20%3D%20%25x%7B%0A%20%20%7B%0A%20%20%20%20a%3A%20%7B%20b%3A%202%20%7D%2C%0A%20%20%20%20c%3A%20%5B%7B%20d%3A%201%20%7D%5D%2C%0A%20%20%7D%0A%7D%0A%0Agon_hash%20%3D%20Hash.new(gon)%0Aa%20%3D%20gon_hash%5B%3Aa%5D%0Ac%20%3D%20gon_hash%5B%3Ac%5D%0A%0Ap%20a%0Ap%20c):

```ruby
gon = %x{
  {
    a: { b: 2 },
    c: [{ d: 1 }],
  }
}

gon_hash = Hash.new(gon)
a = gon_hash[:a]
c = gon_hash[:c]

p a
p c
```

The output is:

```
{"b"=>2}
TypeError: undefined is not a function (evaluating 'i.$__id__()')
$inspect@http://opalrb.org/javascripts/try.js:7:9276
i@http://opalrb.org/javascripts/try.js:5:49029
yield1@http://opalrb.org/javascripts/try.js:4:49598
g@http://opalrb.org/javascripts/try.js:7:4626
http://opalrb.org/javascripts/try.js:5:49052

eval code
eval@[native code]
$eval_code@http://opalrb.org/javascripts/try.js:20:52222
$run_code@http://opalrb.org/javascripts/try.js:20:52122
$@http://opalrb.org/javascripts/try.js:20:77095
o@http://opalrb.org/javascripts/try.js:20:2773
dispatch@http://opalrb.org/javascripts/application.js:3:4887
a@http://opalrb.org/javascripts/application.js:3:780
```

### PR DESCRIPTION
**Pull request by jgaskins**
Fixes #1095

### CODE REVIEW

**Comment by jgaskins** [comment]
Squashing the commits segfaulted the VM. Sweet!

**Comment by elia** [comment]
Restarted the job, I ["gisted" the old log](https://gist.github.com/elia/7c159314ca5e096a30e2) in case anyone wants to submit the crash to the CRuby team :frog:

File: stdlib/native.rb [hunk 1]
```diff
@@ -470,8 +470,16 @@ def initialize(defaults = undefined, &block)
 
           if (value && value.constructor === Object) {
             smap[key] = #{Hash.new(`value`)};
-          }
-          else {
+          } else if (value && value.constructor === Array) {
```
Comment (elia) [inline stdlib/native.rb]: can this be `(value && value.$$is_array)` instead? 

(see https://github.com/opal/opal/blob/master/opal/corelib/array.rb#L8)

**Comment by elia** [comment]
Super happy about this fix, I feel though that it should be extended to arrays in general (vs. just inside `Hash.new`), maybe in `Array.new`. 

Could also make sense to define some kind of general api for classes that, something like `Hash.from_native` and `Array.from_native` which can still be used inside `.new` (eg. for compatibility) but would allow user defined classes to hook in:

```js
if (value && (klass = value.$$class) && klass.$from_native) {
  smap[key] = klass.$from_native(value);
}
```
cc @meh

**Comment by vais** [comment]
@elia :+1: for a separate API for this instead of `native` monkey-patching core classes - I got bitten by this multiple times while working on Opal...

File: stdlib/native.rb (hunk 1 above)
Comment (jgaskins) [inline stdlib/native.rb]: Oh, right, forgot we had those properties to work with. :-)

**Comment by jgaskins** [comment]
I found myself wanting the same thing while writing this. Not being able to do this conversion recursively easily felt a little awkward, and I thought about adding something similar to `from_native`, but I wanted to make a small change before making larger structural changes. :+1:

**Comment by meh** [comment]
A good idea might to be to expand the `Native(..)` functionality and make it use `T.from_native`.

**Comment by elia** [comment]
@vais @jgaskins @meh here's what I'd like to do:

1. merge this PR (after the tiny `.$$is_array` fix)
2. add api docs for native.rb (#1027)
3. start the release cycle for 0.9
2. keep the native.rb changes on master (for the next release)

how does it sound?

**Comment by jgaskins** [comment]
@elia :+1: I'll push the $$is_array thing after I get home.

**Comment by vais** [comment]
Sounds great, @elia!

**Pull request merged**